File: sdk/python/easynet_sdk/health.py
"""Runtime Core health facade."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Mapping, Protocol, runtime_checkable

from .errors import ErrorCode, RetryHint, SDKError, retryable_for_hint


@runtime_checkable
class HealthTransport(Protocol):
    """Narrow transport interface for Runtime Core health."""

    def runtime_health(self) -> bytes:
        """Return raw runtime health JSON bytes from a daemon SDK boundary."""

    def runtime_diagnostics(self) -> bytes:
        """Return raw runtime diagnostics JSON bytes from a daemon SDK boundary."""


@dataclass(frozen=True)
class RuntimeHealth:
    """Language-neutral SDK health DTO."""

    api_ready: bool
    daemon_ready: bool
    invocation_ready: bool
    directory_ready: bool
    trust_ready: bool
    runtime_ready: bool
    version: str | None = None
    abi_version: int | None = None
    mismatch: Mapping[str, object] | None = None
    diagnostics: tuple[str, ...] = field(default_factory=tuple)

    @classmethod
    def from_json(cls, raw: bytes | str) -> "RuntimeHealth":
        """Decode the shared health.schema.json DTO."""

        return _decode_runtime_health(raw)

    def api_alive(self) -> bool:
        """Return process/API liveness, not full runtime readiness."""

        return self.api_ready and self.daemon_ready

    def ready(self) -> bool:
        """Return full runtime readiness."""

        return self.runtime_ready


@dataclass(frozen=True)
class DiagnosticCheck:
    """One named readiness check in a diagnostics report."""

    name: str
    ready: bool
    message: str | None = None


@dataclass(frozen=True)
class DiagnosticsReport:
    """Language-neutral SDK diagnostics DTO."""

    profile: str
    kind: str
    state: str
    ready: bool
    version: str
    abi_version: int
    control_endpoint: str
    invocation_endpoint: str | None
    checks: tuple[DiagnosticCheck, ...]
    diagnostics: tuple[str, ...] = field(default_factory=tuple)

    @classmethod
    def from_json(cls, raw: bytes | str) -> "DiagnosticsReport":
        """Decode the shared diagnostics.schema.json DTO."""

        return _decode_diagnostics_report(raw)


class HealthClient:
    """Python Runtime Core health facade."""

    def __init__(
        self, transport: HealthTransport, *, owns_transport: bool = True
    ) -> None:
        if transport is None:
            raise SDKError(
                code=ErrorCode.INVALID_ARGUMENT,
                stage="sdk",
                retry=RetryHint.NEVER,
                message="health transport is required",
            )
        self._transport = transport
        self._owns_transport = owns_transport
        self._closed = False

    def runtime_health(self) -> RuntimeHealth:
        """Read and decode daemon runtime health."""

        self._require_open()
        try:
            raw = self._transport.runtime_health()
        except SDKError:
            raise
        except Exception as exc:
            raise SDKError(
                code=ErrorCode.ROUTE_UNAVAILABLE,
                stage="transport",
                retry=RetryHint.SAFE,
                retryable=retryable_for_hint(RetryHint.SAFE),
                message="runtime health transport failed",
                cause=exc,
            ) from exc
        return _decode_runtime_health(raw)

    def diagnostics(self) -> DiagnosticsReport:
        """Read and decode daemon runtime diagnostics."""

        self._require_open()
        transport = getattr(self._transport, "runtime_diagnostics", None)
        if transport is None:
            raise SDKError(
                code=ErrorCode.NOT_IMPLEMENTED,
                stage="transport",
                retry=RetryHint.NEVER,
                retryable=False,
                message="health diagnostics transport is not available",
            )
        try:
            raw = transport()
        except SDKError:
            raise
        except Exception as exc:
            raise SDKError(
                code=ErrorCode.ROUTE_UNAVAILABLE,
                stage="transport",
                retry=RetryHint.SAFE,
                retryable=retryable_for_hint(RetryHint.SAFE),
                message="runtime diagnostics transport failed",
                cause=exc,
            ) from exc
        return _decode_diagnostics_report(raw)

    def close(self) -> None:
        """Close the underlying health transport when it owns resources."""

        if self._closed:
            return
        self._closed = True
        close = getattr(self._transport, "close", None)
        if self._owns_transport and close is not None:
            close()

    def _require_open(self) -> None:
        if self._closed:
            raise SDKError(
                code=ErrorCode.INVALID_ARGUMENT,
                stage="sdk",
                retry=RetryHint.NEVER,
                retryable=False,
                message="health client is closed",
            )


def _decode_runtime_health(raw: bytes) -> RuntimeHealth:
    try:
        decoded = json.loads(raw)
    except Exception as exc:
        raise SDKError(
            code=ErrorCode.INVALID_ARGUMENT,
            stage="decode",
            retry=RetryHint.NEVER,
            message=f"decode runtime health JSON: {exc}",
            cause=exc,
        ) from exc
    if not isinstance(decoded, dict):
        raise SDKError(
            code=ErrorCode.INVALID_ARGUMENT,
            stage="decode",
            retry=RetryHint.NEVER,
            message="runtime health JSON must be an object",
        )

    return RuntimeHealth(
        api_ready=_required_bool(decoded, "api_ready"),
        daemon_ready=_required_bool(decoded, "daemon_ready"),
        invocation_ready=_required_bool(decoded, "invocation_ready"),
        directory_ready=_required_bool(decoded, "directory_ready"),
        trust_ready=_required_bool(decoded, "trust_ready"),
        runtime_ready=_required_bool(decoded, "runtime_ready"),
        version=_optional_string(decoded.get("version"), "version"),
        abi_version=_optional_non_negative_int(decoded.get("abi_version"), "abi_version"),
        mismatch=_optional_object(decoded.get("mismatch"), "mismatch"),
        diagnostics=_diagnostics(decoded.get("diagnostics", [])),
    )


def _decode_diagnostics_report(raw: bytes | str) -> DiagnosticsReport:
    try:
        decoded = json.loads(raw)
    except Exception as exc:
        raise SDKError(
            code=ErrorCode.INVALID_ARGUMENT,
            stage="decode",
            retry=RetryHint.NEVER,
            message=f"decode diagnostics JSON: {exc}",
            cause=exc,
        ) from exc
    if not isinstance(decoded, dict):
        raise SDKError(
            code=ErrorCode.INVALID_ARGUMENT,
            stage="decode",
            retry=RetryHint.NEVER,
            message="diagnostics JSON must be an object",
        )
    profile = _required_string(decoded, "profile")
    if profile != "health":
        raise _invalid_health_field("profile", "must be health")
    kind = _required_string(decoded, "kind")
    if kind != "diagnostics_report":
        raise _invalid_health_field("kind", "must be diagnostics_report")
    checks_raw = decoded.get("checks")
    if not isinstance(checks_raw, list) or not checks_raw:
        raise _invalid_health_field("checks", "must be non-empty")
    return DiagnosticsReport(
        profile=profile,
        kind=kind,
        state=_required_string(decoded, "state"),
        ready=_required_bool(decoded, "ready"),
        version=_required_string(decoded, "version"),
        abi_version=_required_non_negative_int(decoded, "abi_version"),
        control_endpoint=_required_string(decoded, "control_endpoint"),
        invocation_endpoint=_optional_string(
            decoded.get("invocation_endpoint"), "invocation_endpoint"
        ),
        checks=tuple(_diagnostic_check(item) for item in checks_raw),
        diagnostics=_diagnostics(decoded.get("diagnostics", [])),
    )


def _diagnostic_check(value: object) -> DiagnosticCheck:
    if not isinstance(value, dict):
        raise _invalid_health_field("checks", "items must be objects")
    return DiagnosticCheck(
        name=_required_string(value, "name"),
        ready=_required_bool(value, "ready"),
        message=_optional_string(value.get("message"), "message"),
    )


def _required_string(decoded: Mapping[str, object], field_name: str) -> str:
    value = decoded.get(field_name)
    if not isinstance(value, str) or value == "":
        raise _invalid_health_field(field_name, "must be a non-empty string")
    return value


def _required_non_negative_int(decoded: Mapping[str, object], field_name: str) -> int:
    value = decoded.get(field_name)
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise _invalid_health_field(field_name, "must be a non-negative integer")
    return value


def _required_bool(decoded: Mapping[str, object], field_name: str) -> bool:
    if field_name not in decoded:
        raise _invalid_health_field(field_name, "is required")
    value = decoded[field_name]
    if not isinstance(value, bool):
        raise _invalid_health_field(field_name, "must be a boolean")
    return value


def _optional_string(value: object, field_name: str) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str):
        raise _invalid_health_field(field_name, "must be a string or null")
    return value


def _optional_non_negative_int(value: object, field_name: str) -> int | None:
    if value is None:
        return None
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise _invalid_health_field(field_name, "must be a non-negative integer or null")
    return value


def _optional_object(value: object, field_name: str) -> Mapping[str, object] | None:
    if value is None:
        return None
    if not isinstance(value, dict):
        raise _invalid_health_field(field_name, "must be an object or null")
    return dict(value)


def _diagnostics(value: object) -> tuple[str, ...]:
    if not isinstance(value, list) or any(not isinstance(item, str) for item in value):
        raise _invalid_health_field("diagnostics", "must be an array of strings")
    return tuple(value)


def _invalid_health_field(field_name: str, message: str) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="decode",
        retry=RetryHint.NEVER,
        message=f"{field_name} {message}",
    )

File: sdk/swift/Sources/EasyNetDaemonSDK/Health.swift
import Foundation

public indirect enum JSONValue: Equatable, Sendable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])
}

public struct RuntimeHealth: Equatable, Sendable {
    public let apiReady: Bool
    public let daemonReady: Bool
    public let invocationReady: Bool
    public let directoryReady: Bool
    public let trustReady: Bool
    public let runtimeReady: Bool
    public let version: String?
    public let abiVersion: Int?
    public let mismatch: [String: JSONValue]?
    public let diagnostics: [String]

    public init(
        apiReady: Bool,
        daemonReady: Bool,
        invocationReady: Bool,
        directoryReady: Bool,
        trustReady: Bool,
        runtimeReady: Bool,
        version: String? = nil,
        abiVersion: Int? = nil,
        mismatch: [String: JSONValue]? = nil,
        diagnostics: [String] = []
    ) {
        self.apiReady = apiReady
        self.daemonReady = daemonReady
        self.invocationReady = invocationReady
        self.directoryReady = directoryReady
        self.trustReady = trustReady
        self.runtimeReady = runtimeReady
        self.version = version
        self.abiVersion = abiVersion
        self.mismatch = mismatch
        self.diagnostics = diagnostics
    }

    public static func fromJSON(_ raw: Data) throws -> RuntimeHealth {
        let object = try decodeObject(raw, label: "runtime health JSON")
        return RuntimeHealth(
            apiReady: try requiredBool(object, "api_ready"),
            daemonReady: try requiredBool(object, "daemon_ready"),
            invocationReady: try requiredBool(object, "invocation_ready"),
            directoryReady: try requiredBool(object, "directory_ready"),
            trustReady: try requiredBool(object, "trust_ready"),
            runtimeReady: try requiredBool(object, "runtime_ready"),
            version: try optionalString(object["version"], "version"),
            abiVersion: try optionalInt(object["abi_version"], "abi_version"),
            mismatch: try optionalObject(object["mismatch"], "mismatch"),
            diagnostics: try diagnosticsList(object["diagnostics"])
        )
    }

    public static func fromJSON(_ raw: String) throws -> RuntimeHealth {
        try fromJSON(Data(raw.utf8))
    }

    public var apiAlive: Bool {
        apiReady && daemonReady
    }

    public var ready: Bool {
        runtimeReady
    }
}

public struct DiagnosticCheck: Equatable, Sendable {
    public let name: String
    public let ready: Bool
    public let message: String?

    public init(name: String, ready: Bool, message: String? = nil) throws {
        guard !name.isEmpty else {
            throw invalidHealthField("checks", "name must be a non-empty string")
        }
        self.name = name
        self.ready = ready
        self.message = message
    }
}

public struct DiagnosticsReport: Equatable, Sendable {
    public let profile: String
    public let kind: String
    public let state: String
    public let ready: Bool
    public let version: String
    public let abiVersion: Int
    public let controlEndpoint: String
    public let invocationEndpoint: String?
    public let checks: [DiagnosticCheck]
    public let diagnostics: [String]

    public init(
        profile: String,
        kind: String,
        state: String,
        ready: Bool,
        version: String,
        abiVersion: Int,
        controlEndpoint: String,
        invocationEndpoint: String?,
        checks: [DiagnosticCheck],
        diagnostics: [String] = []
    ) throws {
        guard profile == "health" else {
            throw invalidHealthField("profile", "must be health")
        }
        guard kind == "diagnostics_report" else {
            throw invalidHealthField("kind", "must be diagnostics_report")
        }
        guard !checks.isEmpty else {
            throw invalidHealthField("checks", "must be non-empty")
        }
        self.profile = profile
        self.kind = kind
        self.state = state
        self.ready = ready
        self.version = version
        self.abiVersion = abiVersion
        self.controlEndpoint = controlEndpoint
        self.invocationEndpoint = invocationEndpoint
        self.checks = checks
        self.diagnostics = diagnostics
    }

    public static func fromJSON(_ raw: Data) throws -> DiagnosticsReport {
        let object = try decodeObject(raw, label: "diagnostics JSON")
        return try DiagnosticsReport(
            profile: requiredString(object, "profile"),
            kind: requiredString(object, "kind"),
            state: requiredString(object, "state"),
            ready: requiredBool(object, "ready"),
            version: requiredString(object, "version"),
            abiVersion: requiredInt(object, "abi_version"),
            controlEndpoint: requiredString(object, "control_endpoint"),
            invocationEndpoint: optionalString(object["invocation_endpoint"], "invocation_endpoint"),
            checks: diagnosticChecks(object["checks"]),
            diagnostics: diagnosticsList(object["diagnostics"])
        )
    }

    public static func fromJSON(_ raw: String) throws -> DiagnosticsReport {
        try fromJSON(Data(raw.utf8))
    }
}

public protocol HealthTransport: AnyObject, Sendable {
    func runtimeHealth() async throws -> Data
    func close() async throws
}

public extension HealthTransport {
    func close() async throws {}
}

public protocol DiagnosticsTransport: AnyObject, Sendable {
    func runtimeDiagnostics() async throws -> Data
}

public final class HealthClient: @unchecked Sendable {
    private let transport: HealthTransport
    private var closed = false

    public init(transport: HealthTransport) {
        self.transport = transport
    }

    public func runtimeHealth() async throws -> RuntimeHealth {
        try requireOpen()
        do {
            return try RuntimeHealth.fromJSON(try await transport.runtimeHealth())
        } catch let error as SDKError {
            throw error
        } catch {
            throw SDKError(
                code: .transport,
                stage: "transport",
                retryHint: .safe,
                retryable: true,
                message: "runtime health transport failed"
            )
        }
    }

    public func diagnostics() async throws -> DiagnosticsReport {
        try requireOpen()
        guard let diagnosticsTransport = transport as? DiagnosticsTransport else {
            throw SDKError(
                code: .notImplemented,
                stage: "transport",
                message: "health diagnostics transport is not available"
            )
        }
        do {
            return try DiagnosticsReport.fromJSON(try await diagnosticsTransport.runtimeDiagnostics())
        } catch let error as SDKError {
            throw error
        } catch {
            throw SDKError(
                code: .transport,
                stage: "transport",
                retryHint: .safe,
                retryable: true,
                message: "runtime diagnostics transport failed"
            )
        }
    }

    public func close() async throws {
        guard !closed else {
            return
        }
        closed = true
        try await transport.close()
    }

    private func requireOpen() throws {
        if closed {
            throw SDKError.closed("health")
        }
    }
}

func decodeObject(_ raw: Data, label: String) throws -> [String: JSONValue] {
    do {
        let decoded = try JSONSerialization.jsonObject(with: raw, options: [])
        guard let object = try jsonValue(decoded).objectValue else {
            throw invalidHealthField(label, "must be an object")
        }
        return object
    } catch let error as SDKError {
        throw error
    } catch {
        throw SDKError(
            code: .invalidArgument,
            stage: "decode",
            message: "decode \(label): \(error)"
        )
    }
}

func jsonValue(_ value: Any) throws -> JSONValue {
    if value is NSNull {
        return .null
    }
    if let bool = value as? Bool {
        return .bool(bool)
    }
    if let number = value as? NSNumber {
        return .number(number.doubleValue)
    }
    if let string = value as? String {
        return .string(string)
    }
    if let array = value as? [Any] {
        return .array(try array.map(jsonValue))
    }
    if let object = value as? [String: Any] {
        return .object(try object.mapValues(jsonValue))
    }
    throw invalidHealthField("json", "contains unsupported value")
}

extension JSONValue {
    var objectValue: [String: JSONValue]? {
        if case let .object(value) = self {
            return value
        }
        return nil
    }
}

private func requiredBool(_ object: [String: JSONValue], _ name: String) throws -> Bool {
    if case let .bool(value) = object[name] {
        return value
    }
    throw invalidHealthField(name, "must be a boolean")
}

private func requiredString(_ object: [String: JSONValue], _ name: String) throws -> String {
    if case let .string(value) = object[name], !value.isEmpty {
        return value
    }
    throw invalidHealthField(name, "must be a non-empty string")
}

private func requiredInt(_ object: [String: JSONValue], _ name: String) throws -> Int {
    if let value = try optionalInt(object[name], name) {
        return value
    }
    throw invalidHealthField(name, "must be a non-negative integer")
}

private func optionalString(_ value: JSONValue?, _ name: String) throws -> String? {
    guard let value else {
        return nil
    }
    switch value {
    case .null:
        return nil
    case let .string(string):
        return string
    default:
        throw invalidHealthField(name, "must be a string or null")
    }
}

private func optionalInt(_ value: JSONValue?, _ name: String) throws -> Int? {
    guard let value else {
        return nil
    }
    switch value {
    case .null:
        return nil
    case let .number(number):
        guard number >= 0, number.rounded() == number, number <= Double(Int.max) else {
            throw invalidHealthField(name, "must be a non-negative integer or null")
        }
        return Int(number)
    default:
        throw invalidHealthField(name, "must be a non-negative integer or null")
    }
}

private func optionalObject(_ value: JSONValue?, _ name: String) throws -> [String: JSONValue]? {
    guard let value else {
        return nil
    }
    switch value {
    case .null:
        return nil
    case let .object(object):
        return object
    default:
        throw invalidHealthField(name, "must be an object or null")
    }
}

private func diagnosticsList(_ value: JSONValue?) throws -> [String] {
    guard let value else {
        return []
    }
    guard case let .array(items) = value else {
        throw invalidHealthField("diagnostics", "must be an array")
    }
    return try items.map { item in
        guard case let .string(diagnostic) = item else {
            throw invalidHealthField("diagnostics", "items must be strings")
        }
        return diagnostic
    }
}

private func diagnosticChecks(_ value: JSONValue?) throws -> [DiagnosticCheck] {
    guard case let .array(items) = value, !items.isEmpty else {
        throw invalidHealthField("checks", "must be non-empty")
    }
    return try items.map { item in
        guard case let .object(check) = item else {
            throw invalidHealthField("checks", "items must be objects")
        }
        return try DiagnosticCheck(
            name: requiredString(check, "name"),
            ready: requiredBool(check, "ready"),
            message: optionalString(check["message"], "message")
        )
    }
}

private func invalidHealthField(_ name: String, _ message: String) -> SDKError {
    SDKError(
        code: .invalidArgument,
        stage: "decode",
        message: "runtime health field \(name) \(message)",
        details: ["field": name]
    )
}
